File: tests/matlab/test_eeg_ms2f.m
function tests = test_eeg_ms2f
% TEST_EEG_MS2F  Tests for eeg_ms2f.m (epoch latency in ms -> nearest 1-based
% epoch frame). outf = 1 + round((pnts-1)*(ms/1000 - xmin)/(xmax - xmin)).
% Closed-form, release-invariant expectations.
tests = functiontests(localfunctions);
end

function eeg = makeEEG(xmin, xmax, pnts)
eeg = struct('xmin', xmin, 'xmax', xmax, 'pnts', pnts);
end

function testFirstFrameAtXmin(testCase)
% ms at xmin -> first frame (1-based).
EEG = makeEEG(0, 1, 1001);
verifyEqual(testCase, eeg_ms2f(EEG, 0), 1);
end

function testLastFrameAtXmax(testCase)
% ms at xmax -> last frame == pnts.
EEG = makeEEG(0, 1, 1001);
verifyEqual(testCase, eeg_ms2f(EEG, 1000), 1001);
end

function testMidpoint(testCase)
% 500 ms -> 0.5 s -> frame 1 + 1000*0.5 = 501.
EEG = makeEEG(0, 1, 1001);
verifyEqual(testCase, eeg_ms2f(EEG, 500), 501);
end

function testRoundsToNearest(testCase)
% 499.4 ms -> 1 + round(499.4) = 500.
EEG = makeEEG(0, 1, 1001);
verifyEqual(testCase, eeg_ms2f(EEG, 499.4), 500);
end

function testEpochCenterNegativeXmin(testCase)
% Epoched data xmin<0: ms=0 -> centre frame.
EEG = makeEEG(-1, 1, 2001);
verifyEqual(testCase, eeg_ms2f(EEG, 0), 1001);
end

function testBelowRangeErrors(testCase)
% ms below xmin -> error('time out of range').
EEG = makeEEG(0, 1, 1001);
threw = false;
try
    eeg_ms2f(EEG, -1);
catch ME
    threw = true;
    verifyTrue(testCase, contains(ME.message, 'out of range'));
end
verifyTrue(testCase, threw);
end

function testAboveRangeErrors(testCase)
% ms above xmax -> error('time out of range').
EEG = makeEEG(0, 1, 1001);
threw = false;
try
    eeg_ms2f(EEG, 2000);
catch ME
    threw = true;
    verifyTrue(testCase, contains(ME.message, 'out of range'));
end
verifyTrue(testCase, threw);
end

File: tests/test_eeg_ms2f.py
"""Parity tests for eeg_ms2f - epoch latency (ms) to nearest epoch frame.

Ports EEGLAB eeg_ms2f.m: ``outf = 1 + round((pnts-1)*(ms/1000 - xmin)/(xmax - xmin))``
with an out-of-range error. Frame numbers are 1-based to match EEGLAB. Expected
values are closed-form and mirror tests/matlab/test_eeg_ms2f.m.
"""

from __future__ import annotations

import pytest

from eegprep.functions.miscfunc.eeg_ms2f import eeg_ms2f

pytestmark = pytest.mark.parity


def _eeg(xmin, xmax, pnts):
    return {"xmin": xmin, "xmax": xmax, "pnts": pnts}


def test_first_frame_at_xmin():
    # ms at xmin -> first frame (1-based).
    assert eeg_ms2f(_eeg(0, 1, 1001), 0) == 1


def test_last_frame_at_xmax():
    # ms at xmax -> last frame == pnts.
    assert eeg_ms2f(_eeg(0, 1, 1001), 1000) == 1001


def test_midpoint():
    # 500 ms -> 0.5 s -> frame 1 + 1000*0.5 = 501.
    assert eeg_ms2f(_eeg(0, 1, 1001), 500) == 501


def test_rounds_to_nearest():
    # 499.4 ms -> 1 + round(499.4) = 500.
    assert eeg_ms2f(_eeg(0, 1, 1001), 499.4) == 500


def test_epoch_center_negative_xmin():
    # Epoched data xmin<0: ms=0 -> centre frame.
    assert eeg_ms2f(_eeg(-1, 1, 2001), 0) == 1001


def test_below_range_raises():
    with pytest.raises(ValueError, match="out of range"):
        eeg_ms2f(_eeg(0, 1, 1001), -1)


def test_above_range_raises():
    with pytest.raises(ValueError, match="out of range"):
        eeg_ms2f(_eeg(0, 1, 1001), 2000)
